cliccl: add ./cockroach mt start-proxy

tbg opened the pull request:
For end-to-end testing purposes, we want a dummy implementation of the
sql proxy in the public repo. This commit adds it; you can now run the
demo below (note the `./cockroach mt start-proxy` part towards the end).
Note that the point is not that this replicates "exactly" the internal
proxy we are planning to deploy, but that it exercises the main
mechanism by which the tenant identifier is read from the connection
parameters sufficiently well to give us confidence that the drivers and
ORMs we test against will work (Examples-ORMs is the main anticipated
user of this).

```bash

set -euxo pipefail

killall -9 cockroach || true

rm -rf certs-{sql,kv,user} cockroach-data
mkdir -p certs-{sql,kv,user}
sql="certs-sql"
kv="certs-kv"

./cockroach cert --certs-dir="${kv}" --ca-key="${kv}/ca.key" create-ca
./cockroach cert --certs-dir="${kv}" --ca-key="${kv}/ca.key" create-node 127.0.0.1 ::1 localhost *.local
./cockroach cert --certs-dir="${kv}" --ca-key="${kv}/ca.key" create-client root

./cockroach mt cert --certs-dir="${kv}" --ca-key="${kv}/ca-server-tenant.key" create-tenant-server-ca
cp "${kv}/ca-server-tenant.crt" "${sql}"
./cockroach mt cert --certs-dir="${kv}" --ca-key="${kv}/ca-server-tenant.key" create-tenant-server 127.0.0.1 ::1 localhost *.local

rm "${kv}"/ca*.key

./cockroach cert --certs-dir="${sql}" --ca-key="${sql}/ca.key" create-ca
./cockroach cert --certs-dir="${sql}" --ca-key="${sql}/ca.key" create-node 127.0.0.1 ::1 localhost *.local

./cockroach cert --certs-dir="${sql}" --ca-key="${sql}/ca.key" create-client root
mv "${sql}"/client.root.* certs-user
cp "${sql}/ca.crt" certs-user

./cockroach mt cert --certs-dir="${sql}" --ca-key="${sql}/ca-client-tenant.key" create-tenant-client-ca
cp "${sql}/ca-client-tenant.crt" "${kv}"
./cockroach mt cert --certs-dir="${sql}" --ca-key="${sql}/ca-client-tenant.key" create-tenant-client 10

rm "${sql}"/ca*.key

./cockroach start-single-node --tenant-addr 127.0.0.1:36257 --certs-dir certs-kv --host 127.0.0.1 --background
./cockroach sql --certs-dir certs-kv --host 127.0.0.1 -e 'select crdb_internal.create_tenant(10);'
rm certs-kv/client.root.*

./cockroach mt start-sql --certs-dir certs-sql --tenant-id 10 --kv-addrs 127.0.0.1:36257 --sql-addr 127.0.0.1:46257 &
sleep 1

./cockroach sql --url 'postgres://root@127.0.0.1:46257?sslmode=verify-ca&sslkey=certs-user/client.root.key&sslcert=certs-user/client.root.crt&sslrootcert=certs-user/ca.crt' -e "CREATE USER bob WITH PASSWORD 'foo';"

./cockroach mt start-proxy --listen-addr 127.0.0.1:56257 --target-addr 127.0.0.1:46257 --listen-cert certs-sql/node.crt --listen-key certs-sql/node.key &
sleep 1

./cockroach sql --url 'postgres://bob:foo@127.0.0.1:56257/prancing-pony.defaultdb?sslmode=verify-ca&sslrootcert=certs-user/ca.crt' -e "SELECT 'this worked!';"

./cockroach sql --url 'postgres://bob:foo@127.0.0.1:56257?options=--cluster=prancing-pony&sslmode=verify-ca&sslrootcert=certs-user/ca.crt'

```

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/52478)
<!-- Reviewable:end -->

tbg commented:
Once merged, this addresses part of https://github.com/cockroachdb/cockroach/issues/52413 - we can update Examples-ORMs to actually use this.

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52478#-:-ME8TiRjE_ceJy0BvSq-:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto and @tbg)

___
*[pkg/ccl/cliccl/mtproxy.go, line 147 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52478#-ME8Swey9l_j8fV-x8NJ:-ME8Swey9l_j8fV-x8NK:b-15cj3d) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dfdf0d373de608d864b9b01ed0d8132347a9981/pkg/ccl/cliccl/mtproxy.go#L147)):*
> ```Go
> 				return sqlProxyTargetAddr, nil
> 			}
> 			if params["options"] == magic {
> ```

Since it's easy enough to do, I think we should just compare against either `"--cluster="+magic` or `"-c cluster="+magic`. There's no reason to let the proxy implementations diverge if avoiding that is straightforward.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52478#-:-ME8aXP1EsA3BxL-SKeG:bg7igkv)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/ccl/cliccl/mtproxy.go, line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52478#-ME8YUrA4buK7h8dW8ra:-ME8YUrA4buK7h8dW8rb:bxqvv2u) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dfdf0d373de608d864b9b01ed0d8132347a9981/pkg/ccl/cliccl/mtproxy.go#L1)):*
> ```Go
> // Copyright 2017 The Cockroach Authors.
> ```

nit: 2020


<!-- Sent from Reviewable.io -->

tbg committed:
cliccl: add ./cockroach mt start-proxy

For end-to-end testing purposes, we want a dummy implementation of the
sql proxy in the public repo. This commit adds it; you can now run the
demo below (note the `./cockroach mt start-proxy` part towards the end).
Note that the point is not that this replicates "exactly" the internal
proxy we are planning to deploy, but that it exercises the main
mechanism by which the tenant identifier is read from the connection
parameters sufficiently well to give us confidence that the drivers and
ORMs we test against will work (Examples-ORMs is the main anticipated
user of this).

```bash

set -euxo pipefail

killall -9 cockroach || true

rm -rf certs-{sql,kv,user} cockroach-data
mkdir -p certs-{sql,kv,user}
sql="certs-sql"
kv="certs-kv"

./cockroach cert --certs-dir="${kv}" --ca-key="${kv}/ca.key" create-ca
./cockroach cert --certs-dir="${kv}" --ca-key="${kv}/ca.key" create-node 127.0.0.1 ::1 localhost *.local
./cockroach cert --certs-dir="${kv}" --ca-key="${kv}/ca.key" create-client root

./cockroach mt cert --certs-dir="${kv}" --ca-key="${kv}/ca-server-tenant.key" create-tenant-server-ca
cp "${kv}/ca-server-tenant.crt" "${sql}"
./cockroach mt cert --certs-dir="${kv}" --ca-key="${kv}/ca-server-tenant.key" create-tenant-server 127.0.0.1 ::1 localhost *.local

rm "${kv}"/ca*.key

./cockroach cert --certs-dir="${sql}" --ca-key="${sql}/ca.key" create-ca
./cockroach cert --certs-dir="${sql}" --ca-key="${sql}/ca.key" create-node 127.0.0.1 ::1 localhost *.local

./cockroach cert --certs-dir="${sql}" --ca-key="${sql}/ca.key" create-client root
mv "${sql}"/client.root.* certs-user
cp "${sql}/ca.crt" certs-user

./cockroach mt cert --certs-dir="${sql}" --ca-key="${sql}/ca-client-tenant.key" create-tenant-client-ca
cp "${sql}/ca-client-tenant.crt" "${kv}"
./cockroach mt cert --certs-dir="${sql}" --ca-key="${sql}/ca-client-tenant.key" create-tenant-client 10

rm "${sql}"/ca*.key

./cockroach start-single-node --tenant-addr 127.0.0.1:36257 --certs-dir certs-kv --host 127.0.0.1 --background
./cockroach sql --certs-dir certs-kv --host 127.0.0.1 -e 'select crdb_internal.create_tenant(10);'
rm certs-kv/client.root.*

./cockroach mt start-sql --certs-dir certs-sql --tenant-id 10 --kv-addrs 127.0.0.1:36257 --sql-addr 127.0.0.1:46257 &
sleep 1

./cockroach sql --url 'postgres://root@127.0.0.1:46257?sslmode=verify-ca&sslkey=certs-user/client.root.key&sslcert=certs-user/client.root.crt&sslrootcert=certs-user/ca.crt' -e "CREATE USER bob WITH PASSWORD 'foo';"

./cockroach mt start-proxy --listen-addr 127.0.0.1:56257 --target-addr 127.0.0.1:46257 --listen-cert certs-sql/node.crt --listen-key certs-sql/node.key &
sleep 1

./cockroach sql --url 'postgres://bob:foo@127.0.0.1:56257/prancing-pony.defaultdb?sslmode=verify-ca&sslrootcert=certs-user/ca.crt' -e "SELECT 'this worked!';"

./cockroach sql --url 'postgres://bob:foo@127.0.0.1:56257?options=--cluster=prancing-pony&sslmode=verify-ca&sslrootcert=certs-user/ca.crt'

```

Release note: None

nvb reviewed (approved):
> Reviewed 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52478)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale)



<!-- Sent from Reviewable.io -->

tbg commented:
bors r=nvanbenschoten,asubiotto
